File: gemini/typescript/examples/resilience.ts
/**
 * Resilience Patterns Example
 *
 * This example demonstrates:
 * - Configuring retry with exponential backoff
 * - Configuring circuit breaker for fault tolerance
 * - Configuring rate limiting
 * - Using ResilienceOrchestrator to combine all patterns
 * - Creating a client with custom resilience configuration
 *
 * Prerequisites:
 * - Set the GEMINI_API_KEY environment variable
 *
 * Usage:
 * ```bash
 * export GEMINI_API_KEY="your-api-key"
 * npm run build
 * node dist/examples/resilience.js
 * ```
 */

import {
  createClient,
  RetryExecutor,
  CircuitBreaker,
  RateLimiter,
  ResilienceOrchestrator,
  CircuitBreakerOpenError,
} from '../src/index.js';

/**
 * Example 1: Retry with exponential backoff
 */
async function exampleRetry(): Promise<void> {
  console.log('\n=== Example 1: Retry with Exponential Backoff ===\n');

  const retry = new RetryExecutor({
    maxAttempts: 3,
    initialDelay: 1000, // 1 second
    maxDelay: 10000, // 10 seconds
    multiplier: 2, // Double delay each time
    jitter: 0.25, // 25% random jitter
  });

  console.log('Retry Configuration:');
  console.log('  Max Attempts: 3');
  console.log('  Initial Delay: 1000ms');
  console.log('  Max Delay: 10000ms');
  console.log('  Multiplier: 2x');
  console.log('  Jitter: 25%\n');

  let attemptCount = 0;

  try {
    const result = await retry.execute(async () => {
      attemptCount++;
      console.log(`Attempt ${attemptCount}...`);

      // Simulate an operation that succeeds on the 2nd attempt
      if (attemptCount < 2) {
        throw new Error('Simulated transient failure');
      }

      return 'Success!';
    });

    console.log(`\nResult: ${result}`);
    console.log(`Total attempts: ${attemptCount}`);
  } catch (error) {
    console.error('Failed after all retries:', error);
  }
}

/**
 * Example 2: Circuit breaker pattern
 */
async function exampleCircuitBreaker(): Promise<void> {
  console.log('\n=== Example 2: Circuit Breaker Pattern ===\n');

  const circuitBreaker = new CircuitBreaker({
    failureThreshold: 3, // Open after 3 failures
    successThreshold: 2, // Close after 2 successes
    openDuration: 5000, // Stay open for 5 seconds
    halfOpenMaxRequests: 1, // Allow 1 request in half-open state
  });

  console.log('Circuit Breaker Configuration:');
  console.log('  Failure Threshold: 3');
  console.log('  Success Threshold: 2');
  console.log('  Open Duration: 5000ms');
  console.log('  Half-Open Max Requests: 1\n');

  // Simulate multiple calls
  for (let i = 1; i <= 7; i++) {
    try {
      console.log(`Call ${i}:`);

      const result = await circuitBreaker.execute(async () => {
        // Simulate failures for first 3 calls
        if (i <= 3) {
          throw new Error('Service unavailable');
        }

        // After circuit opens, wait to let it transition to half-open
        if (i === 4) {
          console.log('  Waiting for circuit to transition to half-open...');
          await new Promise((resolve) => setTimeout(resolve, 5500));
        }

        return `Success ${i}`;
      });

      console.log(`  Result: ${result}`);
      console.log(`  Circuit State: ${circuitBreaker.getState()}`);
    } catch (error) {
      if (error instanceof CircuitBreakerOpenError) {
        console.log(`  Circuit is OPEN - request rejected immediately`);
        console.log(`  Circuit State: ${circuitBreaker.getState()}`);
      } else {
        console.log(`  Error: ${(error as Error).message}`);
        console.log(`  Circuit State: ${circuitBreaker.getState()}`);
      }
    }

    console.log();
  }
}

/**
 * Example 3: Rate limiting
 */
async function exampleRateLimit(): Promise<void> {
  console.log('\n=== Example 3: Rate Limiting ===\n');

  const rateLimiter = new RateLimiter({
    requestsPerMinute: 10,
    tokensPerMinute: 100000,
  });

  console.log('Rate Limiter Configuration:');
  console.log('  Requests per minute: 10');
  console.log('  Tokens per minute: 100,000\n');

  console.log('Making requests...\n');

  // Make 5 rapid requests
  for (let i = 1; i <= 5; i++) {
    const start = Date.now();

    await rateLimiter.execute(async () => {
      const elapsed = Date.now() - start;
      console.log(`Request ${i} executed after ${elapsed}ms`);

      // Simulate token consumption
      rateLimiter.consumeTokens(1000);

      return `Result ${i}`;
    });
  }

  console.log('\nAll requests completed within rate limits!');
}

/**
 * Example 4: ResilienceOrchestrator (combining all patterns)
 */
async function exampleOrchestrator(): Promise<void> {
  console.log('\n=== Example 4: Resilience Orchestrator ===\n');

  const orchestrator = new ResilienceOrchestrator({
    retry: {
      maxAttempts: 3,
      initialDelay: 500,
      maxDelay: 5000,
      multiplier: 2,
      jitter: 0.1,
    },
    circuitBreaker: {
      failureThreshold: 5,
      successThreshold: 2,
      openDuration: 10000,
      halfOpenMaxRequests: 1,
    },
    rateLimit: {
      requestsPerMinute: 60,
      tokensPerMinute: 1000000,
    },
  });

  console.log('Orchestrator combines:');
  console.log('  - Retry with exponential backoff');
  console.log('  - Circuit breaker for fault tolerance');
  console.log('  - Rate limiting for API quotas\n');

  let attemptCount = 0;

  try {
    const result = await orchestrator.execute(async () => {
      attemptCount++;
      console.log(`Orchestrated attempt ${attemptCount}...`);

      // Simulate success on 2nd attempt
      if (attemptCount < 2) {
        throw new Error('Transient error');
      }

      return { success: true, data: 'Operation completed' };
    });

    console.log('\nResult:', result);
    console.log(`Completed after ${attemptCount} attempt(s)`);
  } catch (error) {
    console.error('Operation failed:', error);
  }
}

/**
 * Example 5: Client with custom resilience configuration
 */
async function exampleClientWithResilience(): Promise<void> {
  console.log('\n=== Example 5: Client with Custom Resilience ===\n');

  // Get API key from environment
  const apiKey = process.env.GEMINI_API_KEY;
  if (!apiKey) {
    console.log('GEMINI_API_KEY not set. Skipping this example.');
    return;
  }

  // Create client with custom resilience configuration
  const client = createClient({
    apiKey,
    retry: {
      maxAttempts: 3,
      initialDelay: 1000,
      maxDelay: 30000,
      multiplier: 2,
      jitter: 0.25,
    },
    circuitBreaker: {
      failureThreshold: 5,
      successThreshold: 3,
      openDuration: 30000,
      halfOpenMaxRequests: 1,
    },
    rateLimit: {
      requestsPerMinute: 60,
      tokensPerMinute: 1000000,
    },
  });

  console.log('Client created with resilience configuration:');
  console.log('  Retry: 3 attempts with exponential backoff');
  console.log('  Circuit Breaker: Opens after 5 failures');
  console.log('  Rate Limit: 60 requests/min, 1M tokens/min\n');

  console.log('Making a generation request...');

  try {
    const response = await client.content.generate('gemini-2.0-flash-exp', {
      contents: [
        {
          parts: [{ text: 'Say "Hello, World!" in 5 different languages.' }],
        },
      ],
    });

    if (response.candidates && response.candidates.length > 0) {
      const text = response.candidates[0].content?.parts
        .map((part) => ('text' in part ? part.text : ''))
        .join('');

      console.log('\nResponse:');
      console.log(text);
    }

    console.log('\nRequest completed successfully with resilience patterns applied!');
  } catch (error) {
    console.error('Request failed:', error);
  }
}

/**
 * Example 6: Custom retry logic with selective retries
 */
async function exampleSelectiveRetry(): Promise<void> {
  console.log('\n=== Example 6: Selective Retry Logic ===\n');

  const retry = new RetryExecutor({
    maxAttempts: 3,
    initialDelay: 1000,
    maxDelay: 10000,
    multiplier: 2,
    jitter: 0.1,
  });

  console.log('Demonstrating selective retry based on error type...\n');

  let attemptCount = 0;

  try {
    await retry.execute(async () => {
      attemptCount++;
      console.log(`Attempt ${attemptCount}...`);

      // Simulate different error types
      if (attemptCount === 1) {
        // Retryable: Network error
        const error = new Error('Network timeout');
        (error as any).isRetryable = true;
        throw error;
      } else if (attemptCount === 2) {
        // Not retryable: Validation error
        const error = new Error('Invalid API key');
        (error as any).isRetryable = false;
        throw error;
      }

      return 'Success';
    });
  } catch (error) {
    console.log(`\nFailed with non-retryable error after ${attemptCount} attempts`);
    console.log(`Error: ${(error as Error).message}`);
  }
}

/**
 * Example 7: Monitoring circuit breaker state
 */
async function exampleMonitorCircuitBreaker(): Promise<void> {
  console.log('\n=== Example 7: Monitoring Circuit Breaker State ===\n');

  const circuitBreaker = new CircuitBreaker({
    failureThreshold: 2,
    successThreshold: 2,
    openDuration: 3000,
    halfOpenMaxRequests: 1,
  });

  console.log('Monitoring circuit breaker state transitions...\n');

  // Function to check and display state
  const showState = () => {
    const state = circuitBreaker.getState();
    const stats = circuitBreaker.getStats();
    console.log(`State: ${state}`);
    console.log(`Stats: Successes=${stats.successCount}, Failures=${stats.failureCount}\n`);
  };

  // Initial state
  console.log('Initial state:');
  showState();

  // Cause failures
  console.log('Causing failures...');
  for (let i = 1; i <= 2; i++) {
    try {
      await circuitBreaker.execute(async () => {
        throw new Error('Failure');
      });
    } catch (error) {
      console.log(`Failure ${i} recorded`);
    }
  }
  showState();

  // Try request when open
  console.log('Attempting request when circuit is open...');
  try {
    await circuitBreaker.execute(async () => 'success');
  } catch (error) {
    if (error instanceof CircuitBreakerOpenError) {
      console.log('Request blocked by open circuit');
    }
  }
  showState();

  // Wait for half-open
  console.log('Waiting for circuit to transition to half-open...');
  await new Promise((resolve) => setTimeout(resolve, 3500));
  showState();

  // Successful requests to close circuit
  console.log('Making successful requests...');
  for (let i = 1; i <= 2; i++) {
    await circuitBreaker.execute(async () => {
      console.log(`Success ${i}`);
      return 'success';
    });
  }
  showState();
}

/**
 * Main function running all examples
 */
async function main(): Promise<void> {
  try {
    console.log('=== Resilience Patterns Examples ===');

    await exampleRetry();
    await exampleCircuitBreaker();
    await exampleRateLimit();
    await exampleOrchestrator();
    await exampleClientWithResilience();
    await exampleSelectiveRetry();
    await exampleMonitorCircuitBreaker();

    console.log('\n=== All examples completed successfully! ===');
  } catch (error) {
    console.error('\nError during resilience examples:', error);
    if (error instanceof Error) {
      console.error('Error message:', error.message);
      console.error('Stack trace:', error.stack);
    }
    process.exit(1);
  }
}

// Run the examples
main();

File: gemini/rust/examples/resilience.rs
//! Resilience patterns example for Gemini API.
//!
//! This example demonstrates:
//! - Configuring retry logic with exponential backoff
//! - Configuring circuit breaker patterns
//! - Configuring rate limiting
//! - Using resilience orchestrator
//! - Handling transient failures
//! - Understanding resilience configurations
//!
//! # Usage
//!
//! Set your API key as an environment variable:
//! ```bash
//! export GEMINI_API_KEY="your-api-key-here"
//! ```
//!
//! Then run:
//! ```bash
//! cargo run --example resilience
//! ```

use integrations_gemini::{
    GeminiConfig, GeminiClientImpl,
    resilience::{
        ResilienceOrchestrator, ResilienceConfig,
        RetryConfig, CircuitBreakerConfig, RateLimiterConfig,
        CircuitState,
    },
    types::{Content, Part, Role, GenerateContentRequest},
};
use secrecy::SecretString;
use std::time::Duration;

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    // Initialize tracing
    tracing_subscriber::fmt()
        .with_max_level(tracing::Level::INFO)
        .with_target(false)
        .init();

    println!("=== Gemini Resilience Patterns Example ===\n");

    // Example 1: Default resilience configuration
    println!("=== Example 1: Default Resilience Configuration ===\n");
    example_default_resilience().await?;

    println!("\n");

    // Example 2: Custom retry configuration
    println!("=== Example 2: Custom Retry Configuration ===\n");
    example_custom_retry().await?;

    println!("\n");

    // Example 3: Circuit breaker configuration
    println!("=== Example 3: Circuit Breaker Configuration ===\n");
    example_circuit_breaker().await?;

    println!("\n");

    // Example 4: Rate limiting configuration
    println!("=== Example 4: Rate Limiting Configuration ===\n");
    example_rate_limiting().await?;

    println!("\n");

    // Example 5: Combined resilience patterns
    println!("=== Example 5: Combined Resilience Patterns ===\n");
    example_combined_resilience().await?;

    println!("\n");

    // Example 6: Aggressive configuration
    println!("=== Example 6: Aggressive Configuration (High Reliability) ===\n");
    example_aggressive_resilience().await?;

    println!("\n");

    // Example 7: Conservative configuration
    println!("=== Example 7: Conservative Configuration (Controlled) ===\n");
    example_conservative_resilience().await?;

    println!("\n=== Examples Complete ===");

    Ok(())
}

/// Example: Default resilience configuration.
async fn example_default_resilience() -> Result<(), Box<dyn std::error::Error>> {
    println!("Creating client with default resilience configuration...");

    let api_key = std::env::var("GEMINI_API_KEY")
        .or_else(|_| std::env::var("GOOGLE_API_KEY"))
        .unwrap_or_else(|_| "demo-key".to_string());

    let config = GeminiConfig::builder()
        .api_key(SecretString::new(api_key.into()))
        .build()?;

    let client = GeminiClientImpl::new(config)?;

    println!("\n   ✓ Client created with default resilience settings");
    println!("\n   Default Configuration:");
    println!("     Retry:");
    println!("       • Max attempts: 3");
    println!("       • Initial delay: 1s");
    println!("       • Max delay: 60s");
    println!("       • Multiplier: 2.0");
    println!("       • Jitter: 0.25");
    println!("\n     Circuit Breaker:");
    println!("       • Failure threshold: 5");
    println!("       • Success threshold: 3");
    println!("       • Open duration: 30s");
    println!("       • Half-open max requests: 1");
    println!("\n     Rate Limiter:");
    println!("       • Requests per minute: 60");
    println!("       • Tokens per minute: 1,000,000");

    Ok(())
}

/// Example: Custom retry configuration.
async fn example_custom_retry() -> Result<(), Box<dyn std::error::Error>> {
    println!("Creating orchestrator with custom retry configuration...");

    let retry_config = RetryConfig {
        max_retries: 5,
        initial_delay: Duration::from_millis(500),
        max_delay: Duration::from_secs(30),
        multiplier: 2.0,
        jitter: 0.2,
    };

    let resilience_config = ResilienceConfig {
        retry: retry_config,
        circuit_breaker: CircuitBreakerConfig::default(),
        rate_limiter: RateLimiterConfig::default(),
        enable_retry: true,
        enable_circuit_breaker: true,
        enable_rate_limiting: true,
    };

    let orchestrator = ResilienceOrchestrator::new(resilience_config);

    println!("\n   ✓ Orchestrator created with custom retry settings");
    println!("\n   Custom Retry Configuration:");
    println!("     • Max retries: 5 (instead of default 3)");
    println!("     • Initial delay: 500ms (instead of 1s)");
    println!("     • Max delay: 30s (instead of 60s)");
    println!("     • Multiplier: 2.0 (exponential backoff)");
    println!("     • Jitter: 0.2 (20% randomization to avoid thundering herd)");

    println!("\n   Retry delay progression:");
    let mut delay = 500.0;
    for attempt in 1..=5 {
        println!("     Attempt {}: ~{:.0}ms", attempt + 1, delay);
        delay *= 2.0;
        if delay > 30000.0 {
            delay = 30000.0;
        }
    }

    println!("\n   Use cases:");
    println!("     • More retries for critical operations");
    println!("     • Faster initial retry for transient issues");
    println!("     • Lower max delay for time-sensitive operations");

    Ok(())
}

/// Example: Circuit breaker configuration.
async fn example_circuit_breaker() -> Result<(), Box<dyn std::error::Error>> {
    println!("Creating orchestrator with custom circuit breaker...");

    let circuit_breaker_config = CircuitBreakerConfig {
        failure_threshold: 3,
        success_threshold: 2,
        timeout: Duration::from_secs(60),
        half_open_max_requests: 2,
    };

    let resilience_config = ResilienceConfig {
        retry: RetryConfig::default(),
        circuit_breaker: circuit_breaker_config,
        rate_limiter: RateLimiterConfig::default(),
        enable_retry: true,
        enable_circuit_breaker: true,
        enable_rate_limiting: true,
    };

    let orchestrator = ResilienceOrchestrator::new(resilience_config);

    println!("\n   ✓ Orchestrator created with custom circuit breaker");
    println!("\n   Custom Circuit Breaker Configuration:");
    println!("     • Failure threshold: 3 (open after 3 failures)");
    println!("     • Success threshold: 2 (close after 2 successes in half-open)");
    println!("     • Timeout: 60s (stay open for 60s)");
    println!("     • Half-open max requests: 2 (test with 2 requests)");

    println!("\n   Circuit Breaker States:");
    println!("     1. CLOSED (normal operation)");
    println!("        • Requests pass through normally");
    println!("        • Failures are counted");
    println!("        • Opens after {} failures", 3);
    println!("\n     2. OPEN (failing)");
    println!("        • All requests fail fast");
    println!("        • No requests reach the service");
    println!("        • Transitions to HALF_OPEN after {} seconds", 60);
    println!("\n     3. HALF_OPEN (testing)");
    println!("        • Limited requests pass through (max {})", 2);
    println!("        • Closes after {} successes", 2);
    println!("        • Opens again on any failure");

    // Check current state
    println!("\n   Current circuit state: {:?}", orchestrator.circuit_breaker().state());

    println!("\n   Benefits:");
    println!("     • Prevents cascading failures");
    println!("     • Fast-fail during outages");
    println!("     • Automatic recovery testing");
    println!("     • Protects downstream services");

    Ok(())
}

/// Example: Rate limiting configuration.
async fn example_rate_limiting() -> Result<(), Box<dyn std::error::Error>> {
    println!("Creating orchestrator with custom rate limiting...");

    let rate_limiter_config = RateLimiterConfig {
        requests_per_minute: 30,
        burst_size: 5,
        refill_interval: Duration::from_secs(2),
    };

    let resilience_config = ResilienceConfig {
        retry: RetryConfig::default(),
        circuit_breaker: CircuitBreakerConfig::default(),
        rate_limiter: rate_limiter_config,
        enable_retry: true,
        enable_circuit_breaker: true,
        enable_rate_limiting: true,
    };

    let orchestrator = ResilienceOrchestrator::new(resilience_config);

    println!("\n   ✓ Orchestrator created with custom rate limiting");
    println!("\n   Custom Rate Limiter Configuration:");
    println!("     • Requests per minute: 30");
    println!("     • Burst size: 5 (allow 5 immediate requests)");
    println!("     • Refill interval: 2s (add tokens every 2s)");

    println!("\n   How it works (Token Bucket Algorithm):");
    println!("     1. Bucket starts with 5 tokens (burst size)");
    println!("     2. Each request consumes 1 token");
    println!("     3. Tokens refill every 2 seconds");
    println!("     4. Refill rate: 30 requests/60s = 0.5 requests/s");
    println!("     5. Requests wait if no tokens available");

    println!("\n   Example scenario:");
    println!("     • Send 5 requests immediately (burst) → all succeed");
    println!("     • 6th request → waits ~2s for token refill");
    println!("     • Sustained rate: ~30 requests per minute");

    println!("\n   Use cases:");
    println!("     • Comply with API rate limits");
    println!("     • Prevent overwhelming the service");
    println!("     • Smooth out traffic spikes");
    println!("     • Fair resource allocation");

    Ok(())
}

/// Example: Combined resilience patterns.
async fn example_combined_resilience() -> Result<(), Box<dyn std::error::Error>> {
    println!("Creating orchestrator with all resilience patterns...");

    let resilience_config = ResilienceConfig {
        retry: RetryConfig {
            max_retries: 3,
            initial_delay: Duration::from_millis(1000),
            max_delay: Duration::from_secs(30),
            multiplier: 2.0,
            jitter: 0.25,
        },
        circuit_breaker: CircuitBreakerConfig {
            failure_threshold: 5,
            success_threshold: 3,
            timeout: Duration::from_secs(30),
            half_open_max_requests: 1,
        },
        rate_limiter: RateLimiterConfig {
            requests_per_minute: 60,
            burst_size: 10,
            refill_interval: Duration::from_secs(1),
        },
        enable_retry: true,
        enable_circuit_breaker: true,
        enable_rate_limiting: true,
    };

    let orchestrator = ResilienceOrchestrator::new(resilience_config);

    println!("\n   ✓ Orchestrator created with combined resilience");
    println!("\n   Request Flow:");
    println!("     1. Rate Limiting Check");
    println!("        → Wait if rate limit exceeded");
    println!("\n     2. Circuit Breaker Check");
    println!("        → Fail fast if circuit is open");
    println!("\n     3. Execute Request with Retry");
    println!("        → Retry on transient failures");
    println!("        → Use exponential backoff + jitter");
    println!("\n     4. Update Circuit Breaker");
    println!("        → Record success/failure");
    println!("        → Update circuit state");

    println!("\n   Benefits of Combined Patterns:");
    println!("     • Defense in depth");
    println!("     • Handles different failure modes");
    println!("     • Rate limiting prevents overwhelming");
    println!("     • Circuit breaker provides fast-fail");
    println!("     • Retry handles transient issues");

    Ok(())
}

/// Example: Aggressive configuration for high reliability.
async fn example_aggressive_resilience() -> Result<(), Box<dyn std::error::Error>> {
    println!("Creating orchestrator with aggressive configuration...");

    let resilience_config = ResilienceConfig::aggressive();
    let orchestrator = ResilienceOrchestrator::new(resilience_config);

    println!("\n   ✓ Aggressive configuration applied");
    println!("\n   Aggressive Configuration:");
    println!("     Retry:");
    println!("       • Max retries: 5 (more attempts)");
    println!("       • Initial delay: 500ms (faster retry)");
    println!("       • Max delay: 120s (longer patience)");
    println!("\n     Circuit Breaker:");
    println!("       • Failure threshold: 3 (sensitive)");
    println!("       • Timeout: 60s (longer recovery time)");
    println!("\n     Rate Limiter:");
    println!("       • Requests per minute: 120 (higher throughput)");

    println!("\n   Best for:");
    println!("     • Production systems requiring high availability");
    println!("     • Critical operations that must succeed");
    println!("     • Systems with SLA requirements");
    println!("     • Environments with occasional transient failures");

    Ok(())
}

/// Example: Conservative configuration for controlled environments.
async fn example_conservative_resilience() -> Result<(), Box<dyn std::error::Error>> {
    println!("Creating orchestrator with conservative configuration...");

    let resilience_config = ResilienceConfig::conservative();
    let orchestrator = ResilienceOrchestrator::new(resilience_config);

    println!("\n   ✓ Conservative configuration applied");
    println!("\n   Conservative Configuration:");
    println!("     Retry:");
    println!("       • Max retries: 3 (default)");
    println!("       • Initial delay: 1s (standard)");
    println!("       • Max delay: 60s (reasonable)");
    println!("\n     Circuit Breaker:");
    println!("       • Failure threshold: 10 (lenient)");
    println!("       • Timeout: 30s (standard recovery)");
    println!("\n     Rate Limiter:");
    println!("       • Requests per minute: 30 (controlled)");

    println!("\n   Best for:");
    println!("     • Development and testing");
    println!("     • Systems with strict rate limits");
    println!("     • Environments where failures are rare");
    println!("     • Cost-sensitive applications");

    Ok(())
}
